### DISCUSSION TITLE
mChannel not loaded

### DISCUSSION THREAD

**Discussion by xinyuuu** [assimp/assimp#4148] [Q&A] [answered]
I cannot read the data in aiNodeAnim->mChannel like the position keys all that. It just states that "unable to read the memory". And I only had one mesh but the mNumPositionKeys is like 739095888, same for mNumRotationKeys and ScallingKeys.
By the way, the I'm using .fbx which only consists of one mesh, three animations without bones. So I think it's a vertex animation??

**Comment by Lucipetus** [comment, marked as answer]
It sounds like you need to checkout the mMorphMeshChannels.

**Reply by xinyuuu** [reply]
But the numMorphMeshChannels is 0

**Reply by Lucipetus** [reply]
Well if it contains blend shape animations, that's where they should be found.

You can try set up break points in FBXConverter.cpp, right inside FBXConverter::ConvertAnimationStack(...). You can find the lines where mChannels and mNumChannels are initialized, and continue from there to track down where the data get corrupted.

**Reply by xinyuuu** [reply]
Thank you for your reply. Does it mean that I need to use the FBXConverter as for now I'm just using the importer.ReadFile() to retrieve all the data.  So I am a bit unclear of how to trigger the breakpoint?

**Reply by Lucipetus** [reply]
No, ASSIMP calls its FBXConverter to load FBX files.

**Reply by xinyuuu** [reply]
Thanks. I now can get the channel data but this code (_mScene->mAnimations[i]->mChannels[j]->mNodeName.data != sourceNode->mName.data) always return true. The sourceNode->mName will be RootNode, and then the channel->mNodeName will be Armature or Bone.

**Reply by Lucipetus** [reply]
> 
> 
> Thanks. I now can get the channel data but this code (_mScene->mAnimations[i]->mChannels[j]->mNodeName.data != sourceNode->mName.data) always return true. The sourceNode->mName will be RootNode, and then the channel->mNodeName will be Armature or Bone.

You are comparing the memory addresses of the two. If you want to compare the names, just write "...mNodeName != ...mName".
This will call the overloaded "!=" operator to compare the actual **contents**:
```
    bool operator!=(const aiString &other) const {
        return (length != other.length || 0 != memcmp(data, other.data, length));
    }
```

**Reply by xinyuuu** [reply]
I was initially using the mNodeName != mName but it's the same outcome so I tried using the .data. I'm not sure is it right?
![image](https://user-images.githubusercontent.com/49390657/140287619-72b9cdba-986f-438d-b7a4-7580c10d7ad4.png)

**Reply by Lucipetus** [reply]
Judging by the picture, the nodeName is never the same as the sourceNode's name, so the result is naturally true.
Anyway it's pointless to compare .data.
What do you want to achieve here? Are you searching for the bone node with mNodeName?

**Reply by xinyuuu** [reply]
Yes I want to know which node is affected by the animation so I checked it by comparing their names.
